### ISSUE TITLE
CI wheel building broken due to numpy 1.26 (exotic archs)

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#15333]
### Description

I think the builds have started failing just now, because numpy 1.26 is out. It requires GCC >= 8.4, which the CI setup does not provide.

This failure maybe should make us revisit the idea to compile against a new numpy on an old python... It seems that we have become more sensitive recently to updates to dependencies, which then block every PR.

Sample failing build: https://github.com/astropy/astropy/actions/runs/6212653493/job/16863021788?pr=15304

### Blocked by

* https://github.com/OpenAstronomy/github-actions-workflows/issues/161

**Comment by pllim** [comment]
I asked NumPy people. But in the meantime, we might have to do a `<1.26` pin on `pyproject.toml`.

**Comment by mhvk** [comment]
Thanks for looking into this! Could it be that they don't provide a numpy 1.26 wheel for python 3.9? I agree we probably will need some type of pinning in the interim.

More generally, perhaps we need to fix more (all?) of the build dependencies in the various CI runs...  I feel we're just chasing geese here (with you getting most of the work...).

**Comment by pllim** [comment]
> they don't provide a numpy 1.26 wheel for python 3.9? 

They do at https://pypi.org/project/numpy/#files but I don't know how to translate the filenames to why some builds are failing. Maybe it takes them a while to get all the archs up, or they dropped support for a bunch of stuff? 🤷 

> perhaps we need to fix more (all?) of the build dependencies in the various CI runs.

I think it is quite unreasonable to ask us to *build* with unstable numpy. We already test against dev numpy at runtime and none of these were caught. Ideas are welcome though.

**Comment by mhvk** [comment]
I don't think `numpy 1.26` is itself unstable, it just requires some relatively new environment to be built in...

My suggestion: we fix the build environment (i.e., all versions fixed) any time we have a new release.

But right now your suggestion of a numpy pin seems the easiest! It could be for the wheel build one only 

p.s. You probably already checked, but nominally numpy 1.26 supports python 3.9, so indeed I'd think there be wheels for it. It may well be that the problem is `pip/pypi` here, going for a source build when we'd happily take numpy 1.25. No idea whether one can turn that off...

**Comment by pllim** [comment]
> fix the build environment

I don't understand what you mean here. The build environments seem fine to me until numpy 1.26 came out. What needs fixing?

**Comment by pllim** [comment]
> I don't think `numpy 1.26` is itself unstable

To clarify, I was referring to future-proofing our own builds against unreleased numpy, which I think is an unwinnable battle, especially since they might not even push dev wheels for all the archs we test against.

**Comment by pllim** [comment]
Numpy also said something about we have to "update our environments" and pointed to https://docs.scipy.org/doc/scipy/dev/toolchain.html but I don't have time to look at this. So if someone is passionate about builds, please propose a better fix than #15336 . Thanks!

**Comment by pllim** [comment]
p.s. Since we also grab workflows from OpenAstronomy, I don't even know if it is something we can fix here or someone needs to patch https://github.com/OpenAstronomy/github-actions-workflows/blob/main/.github/workflows/publish.yml .

@astrofrog @ConorMacBride ?

**Comment by ngoldbaum** [comment]
In particular, the wheel building emvironment needs to be updates to the manylinux2 build images for wheels that depend on numpy 1.26 or newer.

**Comment by pllim** [comment]
Yes, I agree. https://github.com/astropy/astropy/pull/15336 is a no-go.

**Comment by pllim** [comment]
For those who read this issue later, while this does make a mess of the CI, astropy at runtime for typical users should not be affected. We already tested against unreleased numpy and 1.26rc1, so I have high confidence the code would work. This should only affect the *builds* (which would theoretically block our releases if not fixed quickly enough).

**Comment by pllim** [comment]
I misunderstood the problem. William pointed out that we might indeed be able to fix within this repo. Thanks, William!

**Comment by astrofrog** [comment]
> In particular, the wheel building environment needs to be updated to the manylinux2 build images for wheels that depend on numpy 1.26 or newer.

I'm pretty sure we can configure this via pyproject.toml - will look shortly

**Comment by pllim** [comment]
Update: The normal wheels are fixed at  #15339 (thanks, William)!

There is still the cron job with "exotic" archs that is failing with 3/4 archs. We do not use cibuildwheel for those, so they have to be investigated separately.

tl;dr -- PRs with "Extra CI" label are still blocked.

**Comment by WilliamJamieson** [comment]
#15339 resolved the main problem raised in this issue. The exotic archs CI is really a separate issue from this discussion, as such I am closing this issue in favor of #15351 which is specifically focused on that problem.
